db/migrate/20220401071321_create_context_notifications.rb

```diff
@@ -0,0 +1,13 @@
+class CreateContextNotifications < ActiveRecord::Migration[7.0]
+  def change
+    create_table :context_notifications do |t|
+      t.string :action
+      t.integer :context_id
```

Comment:
It makes sense though I supposed that we were going to query mostly by `context_id` and `context_type` along with `action`, at least for now, so I thought that the combined index could be enough.

Reply:
@jeremyf This is a really great opportunity to highlight another awesome feature of Postgres: The query planner can use composite `BTREE` indexes (the default index type and the one we mean implicitly when we talk about indexes in a DB) even when not all columns/expressions appear in the query. From [the documentation](https://www.postgresql.org/docs/current/indexes-multicolumn.html):

<img width="853" alt="A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns" src="https://user-images.githubusercontent.com/108205/163404930-439e21c2-3598-4ef6-af8c-2372e44bc8b0.png">

More concretely: @lightalloy's got us covered with the unique index on line 10-11 as long as `context_id`, `context_type`, _or_ `action` appear in the `WHERE` clause. 👍🏼

---

💡 Fun fact, unique indexes are pretty much the only time you want to use multicolumn `BTREE` indexes because Postgres can also combine multiple single-column indexes, which take up less space.

There's some more info about this on the doc page I linked above. It's not super intuitive so very few people know about it, so I end up getting to teach people about it everywhere I work.

Reply:
I think I knew about this but never quite internalized it.  Thanks for taking the time to point it out.